extractor/src/main/scala/org/jetbrains/sbt/extractors/DependenciesExtractor.scala

```diff
@@ -95,19 +154,163 @@ object DependenciesExtractor extends SbtStateOps with TaskOps {
+        val projectDependencies = retrieveTransitiveProjectDependencies(projectRef, config, settings, buildDependencies)
+        (Configuration(selfConfig.name), projectDependencies)
+      }
+
+    val allConfigurationsWithSourceOfAllProjectsNames = allConfigurationsWithSourceOfAllProjects.map(_.name)
+    val configToProjects: Map[Configuration, Seq[ProjectRef]] = configToDependencies.mapValues {
+        keepProjectsWithAtLeastOneSourceConfig(_, allConfigurationsWithSourceOfAllProjectsNames)
+      }
+
+    invert(configToProjects)
+  }
+
+  /**
+   * Retrieving project transitive dependencies is done using [[sbt.Classpaths.interSort]] from sbt.
+   * It finds all transitive dependencies by static analysis of the configuration strings (like e.g. "compile" or "compile->test").
+   * The dependencies returned from [[sbt.Classpaths.interSort]] are increased by the dependencies with configurations
+   * that a target configuration inherits from.
+   * So for such project: {{{
+   *   val root = project.dependsOn(
+   *     proj1 % "compile->test"
+   *   )
+   * }}}
+   * [[sbt.Classpaths.interSort]] for `compile` scope in project `root` will return: {{{
+   *   compile -> (proj1, "compile"), (proj1, "test"), (proj1, "runtime")
+   * }}}
+   * It will return such a result because `test` configuration extends `runtime` and `runtime` configuration extends `compile`
+   */
+  private def retrieveTransitiveProjectDependencies(
+    projectRef: ProjectRef,
+    config: sbt.Configuration,
+    settings: Settings[Scope],
+    buildDependencies: BuildDependencies
+  ): Seq[ProjectDependency] = {
+    val allDependencies = Classpaths.interSort(projectRef, config, settings, buildDependencies)
+    // note: when production and test sources will be separated removing all dependencies
+    // with origin project itself, should be done more carefully because it will be required to put projectRef production sources in test sources
+    val dependenciesWithoutProjectItself = allDependencies.filter(_._1 != projectRef)
+    dependenciesWithoutProjectItself.map(ProjectDependency.apply)
+  }
+
+  /**
+   * The goal of this method is to remove project dependencies with configurations that do not have sources e.g. provided.
+   * It needs to be done because [[sbt.Classpaths.interSort]] generate transitive dependencies by analyzing configurations strings (e.g. "compile->provided")
+   * and not every configuration string means that this dependency should actually be added.
+   * So for such project: {{{
+   *   val root = project.dependsOn(
+   *     proj1 % "compile->provided",
+   *     proj2 % "compile"
+   *   )
+   * }}}
+   * [[sbt.Classpaths.interSort]] for `compile` scope in project `root` will return: {{{
+   *   compile -> (proj1, "provided"), (proj2, "compile")
+   * }}}
+   * which in practice means that we only have to add `proj2` as a dependency to `root` and `proj1` dependency shouldn't be taken into account.
```

Comment:
tested in `testNonSourceConfigurationsWithNestedProjectDependencies` in scala plugin